# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/vm/AtomsTable.h
# Commit: e85578d93db4
# Full Hash: e85578d93db4c1060065cfbeffe849d414fcf9d4
# Author: Jon Coppeard <jcoppeard@mozilla.com>
# Date: 2019-05-01 15:57:23
# Regressor Bug: 1547565
# File Overlap Count: 1
# Description:
#   Bug 1547565 - Remove AtomStateEntry and use ReadBarriered<JSAtom*> instead in the atoms table r=jandem
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D29179
# ==============================================================================

diff -r 20ca9c77f630 -r e85578d93db4 js/src/vm/AtomsTable.h
--- a/js/src/vm/AtomsTable.h	Mon Apr 29 16:06:09 2019 +0100
+++ b/js/src/vm/AtomsTable.h	Mon Apr 29 11:09:33 2019 +0100
@@ -39,53 +39,15 @@
   ~AutoLockAllAtoms();
 };
 
-class AtomStateEntry {
-  uintptr_t bits;
-
-  static const uintptr_t NO_TAG_MASK = uintptr_t(-1) - 1;
-
- public:
-  AtomStateEntry() : bits(0) {}
-  AtomStateEntry(const AtomStateEntry& other) : bits(other.bits) {}
-  AtomStateEntry(JSAtom* ptr, bool tagged)
-      : bits(uintptr_t(ptr) | uintptr_t(tagged)) {
-    MOZ_ASSERT((uintptr_t(ptr) & 0x1) == 0);
-  }
-
-  bool isPinned() const { return bits & 0x1; }
-
-  /*
-   * Non-branching code sequence. Note that the const_cast is safe because
-   * the hash function doesn't consider the tag to be a portion of the key.
-   */
-  void setPinned(bool pinned) const {
-    const_cast<AtomStateEntry*>(this)->bits |= uintptr_t(pinned);
-  }
-
-  JSAtom* asPtrUnbarriered() const {
-    MOZ_ASSERT(bits);
-    return reinterpret_cast<JSAtom*>(bits & NO_TAG_MASK);
-  }
-
-  JSAtom* asPtr(JSContext* cx) const;
-
-  bool needsSweep() {
-    JSAtom* atom = asPtrUnbarriered();
-    return gc::IsAboutToBeFinalizedUnbarriered(&atom);
-  }
-};
-
 struct AtomHasher {
   struct Lookup;
   static inline HashNumber hash(const Lookup& l);
-  static MOZ_ALWAYS_INLINE bool match(const AtomStateEntry& entry,
+  static MOZ_ALWAYS_INLINE bool match(const ReadBarriered<JSAtom*>& entry,
                                       const Lookup& lookup);
-  static void rekey(AtomStateEntry& k, const AtomStateEntry& newKey) {
-    k = newKey;
-  }
 };
 
-using AtomSet = JS::GCHashSet<AtomStateEntry, AtomHasher, SystemAllocPolicy>;
+using AtomSet = JS::GCHashSet<ReadBarriered<JSAtom*>, AtomHasher,
+                              SystemAllocPolicy>;
 
 // This class is a wrapper for AtomSet that is used to ensure the AtomSet is
 // not modified. It should only expose read-only methods from AtomSet.